cardano-cli/src/Cardano/CLI/EraBased/HashChecking.hs

```diff
@@ -0,0 +1,123 @@
+  , checkVotingProcedureHashes
+  , checkProposalHashes
+  )
+where
+
+import           Cardano.Api (Certificate (..), ExceptT, firstExceptT)
+import qualified Cardano.Api.Ledger as L
+import qualified Cardano.Api.Shelley as Shelley
+
+import           Cardano.CLI.Run.Hash (carryHashChecks)
+import           Cardano.CLI.Types.Common (MustCheckHash (..), PotentiallyCheckedAnchor (..))
+import           Cardano.CLI.Types.Errors.TxCmdError (TxCmdError (..))
+import qualified Cardano.Ledger.Api.Governance as L
+
+import           Control.Monad (forM_)
+import           Control.Monad.Trans.Except.Extra (left)
+
+-- | Check the hash of the anchor data against the hash in the anchor
+checkAnchorMetadataHash :: L.Anchor L.StandardCrypto -> ExceptT TxCmdError IO ()
+checkAnchorMetadataHash anchor =
+  firstExceptT (TxCmdHashCheckError $ L.anchorUrl anchor) $
+    carryHashChecks
+      ( PotentiallyCheckedAnchor
+          { pcaMustCheck = CheckHash
+          , pcaAnchor = anchor
+          }
+      )
+
+-- | Find references to anchor data and check the hashes are valid
+-- and they match the linked data.
+checkCertificateHashes :: Certificate era -> ExceptT TxCmdError IO ()
+checkCertificateHashes c =
+  case c of
+    ShelleyRelatedCertificate _ shelleyCert ->
+      case shelleyCert of
+        L.ShelleyTxCertDelegCert shelleyDelegCert ->
+          case shelleyDelegCert of
+            L.ShelleyRegCert _ -> return ()
+            L.ShelleyUnRegCert _ -> return ()
+            L.ShelleyDelegCert _ _ -> return ()
+        L.ShelleyTxCertPool shelleyPoolCert ->
+          case shelleyPoolCert of
+            L.RegPool poolParams -> forM_ (L.ppMetadata poolParams) checkPoolMetadataHash
+            L.RetirePool _ _ -> return ()
+        L.ShelleyTxCertGenesisDeleg _ -> return ()
+        L.ShelleyTxCertMir _ -> return ()
+    ConwayCertificate ceo conwayCert ->
+      Shelley.conwayEraOnwardsConstraints ceo $
+        case conwayCert of
+          L.ConwayTxCertDeleg _ -> return ()
+          L.ConwayTxCertPool conwayPoolCert ->
+            case conwayPoolCert of
+              L.RegPool poolParams -> forM_ (L.ppMetadata poolParams) checkPoolMetadataHash
+              L.RetirePool _ _ -> return ()
+          L.ConwayTxCertGov govCert ->
+            case govCert of
+              L.ConwayRegDRep _ _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+              L.ConwayUnRegDRep _ _ -> return ()
+              L.ConwayUpdateDRep _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+              L.ConwayAuthCommitteeHotKey _ _ -> return ()
+              L.ConwayResignCommitteeColdKey _ mAnchor -> forM_ mAnchor checkAnchorMetadataHash
+ where
+  checkPoolMetadataHash :: L.PoolMetadata -> ExceptT TxCmdError IO ()
+  checkPoolMetadataHash (L.PoolMetadata{L.pmUrl = url, L.pmHash = hashBytes}) = do
+    let mHash = L.hashFromBytes hashBytes
+    hash <- maybe (left $ TxCmdPoolMetadataHashError url) return mHash
+    let safeHash = L.unsafeMakeSafeHash hash
```

Comment:
>-- | Don't use this except in Testing to make Arbitrary instances, etc.
>--   Defined here, only because the Constructor is in scope here.

This comment on `L.unsafeMakeSafeHash` is a bit worrying. But I see that ledger does not expose a nice way of getting the pool hash.

Maybe we should provide an instance:
```
instance HashWithCrypto PoolMetadata PoolMetadata
```
to be able to calculate the hash the right way? I'm guessing here.

This hash calculation logic should be implemented as a function in `cardano-api` I think. Maybe even upstreamed to ledger?

Reply:
Yes, I think `SafeHash` is widely misused (doesn't really have any guarantees anymore). The fact that there are parsers that produce `SafeHash`. That makes no sense to me, because there is no way they check the hash corresponds to something. But some functions do require `SafeHash` for some reason, when they shouldn't, so there is no way around it.

Reply:
The idea is `SafeHash` must internally store the data that hashes to that (as I understand it), to ensure it is a hash of something. But here we are reading the actual hash (not the thing that is hashed) from command line, so we cannot get that info. But we do need a `SafeHash` 🤷‍♂️